qutebrowser/browser/webengine/spell.py

```diff
@@ -0,0 +1,142 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et
+
+# Copyright 2017 Michal Siedlaczek <michal.siedlaczek@gmail.com>
+
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Installing and configuring spell-checking for QtWebEngine."""
+
+import os
+from urllib.request import urlretrieve
+
+from PyQt5.QtCore import QLibraryInfo
+
+
+class Language:
+
+    """Dictionary language specs."""
+
+    def __init__(self, code, name, file):
+        self.code = code
+        self.name = name
+        self.file = file
+
+    @staticmethod
+    def from_array(lang_array):
+        """Create Language object from an array.
+
+        Args:
+            lang_array: an array of strings containing
+                        the specs of the language in the following format:
+                        [code, name, file]
+        """
+        return Language.from_tuple(tuple(lang_array))
+
+    @staticmethod
+    def from_tuple(lang_tuple):
+        """Create Language object from a tuple.
+
+        Args:
+            lang_tuple: a tuple of strings containing
+                        the specs of the language in the following format:
+                        (code, name, file)
+        """
+        code, name, file = lang_tuple
+        return Language(code, name, file)
+
+    @staticmethod
+    def from_tsv_string(tsv_string):
+        """Create Language object from a string in tab-separated values format.
+
+        Args:
+            tsv_string: a string containing
+                        the specs of the language in the following format:
+                        "code   name    file"
+        """
+        lang_array = tsv_string.split('\t')
+        return Language.from_array(lang_array)
+
+    def __repr__(self):
+        return 'Language({}, {}, {})'.format(self.code, self.name, self.file)
+
+
+def get_dictionary_dir():
+    """Return the path to the QtWebEngine's dictionaries directory."""
+    return QLibraryInfo.location(QLibraryInfo.DataPath) + \
+        '/qtwebengine_dictionaries'
+
+
+def get_language_list_file():
+    """Return the path to the file with the list of all available languages."""
+    root_dir = os.path.dirname(os.path.abspath(__file__))
+    # TODO: not sure how to determine the following path 'the right way'
```

Comment:
Not sure if this is the 'right way', but there's a [basedir](https://github.com/qutebrowser/qutebrowser/blob/master/qutebrowser/__init__.py) var in the root. I'm not sure if you can dereference it directly or not though.